server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -0,0 +1,366 @@
+import java.nio.file.NoSuchFileException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * A RemoteDirectory extension for remote segment store. We need to make sure we don't overwrite a segment file once uploaded.
+ * In order to prevent segment overwrite which can occur due to two primary nodes for the same shard at the same time,
+ * a unique suffix is added to the uploaded segment file. This class keeps track of filename of segments stored
+ * in remote segment store vs filename in local filesystem and provides the consistent Directory interface so that
+ * caller will be accessing segment files in the same way as {@code FSDirectory}. Apart from storing actual segment files,
+ * remote segment store also keeps track of refresh checkpoints as metadata in a separate path which is handled by
+ * another instance of {@code RemoteDirectory}.
+ * @opensearch.internal
+ */
+public final class RemoteSegmentStoreDirectory extends FilterDirectory {
+    /**
+     * Each segment file is uploaded with unique suffix.
+     * For example, _0.cfe in local filesystem will be uploaded to remote segment store as _0.cfe__gX7bNIIBrs0AUNsR2yEG
+     */
+    public static final String SEGMENT_NAME_UUID_SEPARATOR = "__";
+
+    public static final MetadataFilenameUtils.MetadataFilenameComparator METADATA_FILENAME_COMPARATOR =
+        new MetadataFilenameUtils.MetadataFilenameComparator();
+
+    /**
+     * remoteDataDirectory is used to store segment files at path: cluster_UUID/index_UUID/shardId/segments/data
+     */
+    private final RemoteDirectory remoteDataDirectory;
+    /**
+     * remoteMetadataDirectory is used to store metadata files at path: cluster_UUID/index_UUID/shardId/segments/metadata
+     */
+    private final RemoteDirectory remoteMetadataDirectory;
+
+    /**
+     * To prevent explosion of refresh metadata files, we replace refresh files for the given primary term and generation
+     * This is achieved by uploading refresh metadata file with the same UUID suffix.
+     */
+    private String metadataFileUniqueSuffix;
+
+    /**
+     * Keeps track of local segment filename to uploaded filename along with other attributes like checksum.
+     * This map acts as a cache layer for uploaded segment filenames which helps avoid calling listAll() each time.
+     * It is important to initialize this map on creation of RemoteSegmentStoreDirectory and update it on each upload and delete.
+     */
+    private Map<String, UploadedSegmentMetadata> segmentsUploadedToRemoteStore;
+
+    private static final Logger logger = LogManager.getLogger(RemoteSegmentStoreDirectory.class);
+
+    public RemoteSegmentStoreDirectory(RemoteDirectory remoteDataDirectory, RemoteDirectory remoteMetadataDirectory) throws IOException {
```

Comment:
The only concern I see with having two RemoteDirectory instances is they can fundamentally be two separate stores altogether with varying consistency properties. Do we really need them to be exposed separately. Can we not keep the metadata repository internal to this class, to avoid possible misuse of this class.

Reply:
`remoteMetadataDirectory` will be internal to this class. Callers of `RemoteSegmentStoreDirectory` will not be aware of the existence of two different remote directories. 

On that part of creating an instance of `RemoteSegmentStoreDirectory`, it will be handled by the factory. The implementation of factory method will be part of subsequent PR but you can find the code in this commit: https://github.com/sachinpkale/OpenSearch/blob/remote_segment_failure_handling/server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectoryFactory.java

Reply:
Sure but the uploadMetadata is still public
```
 public void uploadMetadata(Collection<String> segmentFiles, Directory storeDirectory, long primaryTerm, long generation)
```

Reply:
Yes, that is the plan. I haven't made that change as not sure on how to decide the trigger for metadata upload. 
Today, it will be called by `RemoteStoreRefereshListener` as `RemoteStoreRefreshListener` knows when the refresh is happening.

Reply:
@Bukhtawar Created a tracking issue for the same: https://github.com/opensearch-project/OpenSearch/issues/4171